File: leetcode/78. Subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> curr;
        vector<vector<int>> res;
        helper(res, curr, nums, 0);
        return res;
    }
private:
    void helper(vector<vector<int>>& res, vector<int>& curr, vector<int>& nums, int index){
        res.push_back(curr);

        for(int i = index; i < nums.size(); ++i){
            curr.push_back(nums[i]);
            helper(res, curr, nums, i + 1);
            curr.pop_back();
        }
    }
};

File: leetcode/78. Subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        def helper(start, curr):
            res.append(curr[:])
            for i in range(start, len(nums)):
                curr.append(nums[i])
                helper(i+1, curr)
                curr.pop()
        helper(0, [])
        return res


class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = [[]]
        for num in nums:
            sub = []
            for lst in res:
                tmp = lst[:]
                tmp.append(num)
                sub.append(tmp)
            res += sub
        return res
